File: Services/GardenService.swift
import Foundation
import SwiftData

struct GardenService {

    // MARK: - Seed Operations
    // These deterministic operations are the target for future voice interaction.
    // The GardenAssistant layer will invoke these rather than simulating UI taps.

    static func addSeed(
        cropName: String,
        variety: String? = nil,
        state: SeedState = .own,
        notes: String? = nil,
        garden: Garden? = nil,
        in context: ModelContext
    ) -> Seed {
        let seed = Seed(cropName: cropName, variety: variety, state: state, notes: notes)
        seed.garden = garden
        context.insert(seed)
        return seed
    }

    static func updateSeed(
        _ seed: Seed,
        cropName: String,
        variety: String?,
        state: SeedState,
        notes: String?
    ) {
        seed.cropName = cropName
        seed.variety = variety
        seed.state = state
        seed.notes = notes
        seed.updatedAt = Date()
    }

    static func deleteSeed(_ seed: Seed, in context: ModelContext) {
        context.delete(seed)
    }

    static func changeSeedState(_ seed: Seed, to newState: SeedState) {
        seed.state = newState
        seed.updatedAt = Date()
    }

    // MARK: - Growing Space Operations

    static func addGrowingSpace(
        name: String,
        notes: String? = nil,
        garden: Garden? = nil,
        in context: ModelContext
    ) -> GrowingSpace {
        let space = GrowingSpace(name: name, notes: notes)
        space.garden = garden
        context.insert(space)
        return space
    }

    static func updateGrowingSpace(
        _ space: GrowingSpace,
        name: String,
        notes: String?
    ) {
        space.name = name
        space.notes = notes
        space.updatedAt = Date()
    }

    static func deleteGrowingSpace(_ space: GrowingSpace, in context: ModelContext) {
        context.delete(space)
    }

    // MARK: - Occupancy Operations

    static func recordPlanting(
        cropName: String,
        variety: String? = nil,
        startDate: Date,
        growingSpace: GrowingSpace? = nil,
        plant: Plant? = nil,
        expectedHarvestDate: Date? = nil,
        expectedReleaseDate: Date? = nil,
        notes: String? = nil,
        in context: ModelContext
    ) -> Occupancy {
        let occupancy = Occupancy(
            cropName: cropName,
            variety: variety,
            startDate: startDate,
            expectedHarvestDate: expectedHarvestDate,
            expectedReleaseDate: expectedReleaseDate,
            status: .active,
            notes: notes,
            growingSpace: growingSpace,
            plant: plant
        )
        if let plant {
            plant.occupancy = occupancy
        }
        context.insert(occupancy)
        return occupancy
    }

    static func completeOccupancy(_ occupancy: Occupancy, endDate: Date? = nil) {
        occupancy.endDate = endDate ?? Date()
        occupancy.status = .completed
        occupancy.updatedAt = Date()
    }

    static func updateOccupancy(
        _ occupancy: Occupancy,
        cropName: String,
        variety: String?,
        startDate: Date,
        expectedHarvestDate: Date?,
        expectedReleaseDate: Date?,
        notes: String?
    ) {
        occupancy.cropName = cropName
        occupancy.variety = variety
        occupancy.startDate = startDate
        occupancy.expectedHarvestDate = expectedHarvestDate
        occupancy.expectedReleaseDate = expectedReleaseDate
        occupancy.notes = notes
        occupancy.updatedAt = Date()
    }

    static func deleteOccupancy(_ occupancy: Occupancy, in context: ModelContext) {
        context.delete(occupancy)
    }

    // MARK: - Desire Operations

    static func createDesire(
        cropName: String,
        variety: String? = nil,
        notes: String? = nil,
        garden: Garden? = nil,
        in context: ModelContext
    ) -> Desire {
        let desire = Desire(cropName: cropName, variety: variety, notes: notes, garden: garden)
        context.insert(desire)
        return desire
    }

    static func updateDesire(
        _ desire: Desire,
        cropName: String,
        variety: String?,
        notes: String?
    ) {
        desire.cropName = cropName
        desire.variety = variety
        desire.notes = notes
        desire.updatedAt = Date()
    }

    static func fulfillDesire(_ desire: Desire) {
        desire.isFulfilled = true
        desire.isCancelled = false
        desire.isExpired = false
        desire.updatedAt = Date()
    }

    static func cancelDesire(_ desire: Desire) {
        desire.isCancelled = true
        desire.isFulfilled = false
        desire.isExpired = false
        desire.updatedAt = Date()
    }

    static func expireDesire(_ desire: Desire) {
        desire.isExpired = true
        desire.isFulfilled = false
        desire.isCancelled = false
        desire.updatedAt = Date()
    }

    static func deleteDesire(_ desire: Desire, in context: ModelContext) {
        context.delete(desire)
    }
}

File: app/src/main/java/com/soilandsupper/service/GardenService.kt
package com.soilandsupper.service

import com.soilandsupper.domain.model.Desire
import com.soilandsupper.domain.model.Garden
import com.soilandsupper.domain.model.GrowingSpace
import com.soilandsupper.domain.model.Occupancy
import com.soilandsupper.domain.model.OccupancyStatus
import com.soilandsupper.domain.model.PlannedPlanting
import com.soilandsupper.domain.model.Plant
import com.soilandsupper.domain.model.Seed
import com.soilandsupper.domain.model.SeedState

object GardenService {

    fun addSeed(
        cropName: String,
        variety: String? = null,
        state: SeedState = SeedState.OWN,
        notes: String? = null,
        garden: Garden? = null
    ): Seed {
        return Seed(
            cropName = cropName,
            variety = variety,
            state = state.name,
            notes = notes,
            gardenId = garden?.id
        )
    }

    fun updateSeed(
        seed: Seed,
        cropName: String,
        variety: String?,
        state: SeedState,
        notes: String?
    ): Seed {
        return seed.copy(
            cropName = cropName,
            variety = variety,
            state = state.name,
            notes = notes,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun addGrowingSpace(
        name: String,
        notes: String? = null,
        spaceType: String? = null,
        width: Double? = null,
        length: Double? = null,
        garden: Garden? = null
    ): GrowingSpace {
        return GrowingSpace(
            name = name,
            notes = notes,
            spaceType = spaceType,
            width = width,
            length = length,
            gardenId = garden?.id
        )
    }

    fun updateGrowingSpace(
        space: GrowingSpace,
        name: String,
        notes: String?,
        spaceType: String?,
        width: Double?,
        length: Double?
    ): GrowingSpace {
        return space.copy(
            name = name,
            notes = notes,
            spaceType = spaceType,
            width = width,
            length = length,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun recordPlanting(
        cropName: String,
        variety: String? = null,
        startDate: Long,
        growingSpace: GrowingSpace? = null,
        plant: Plant? = null,
        expectedHarvestDate: Long? = null,
        expectedReleaseDate: Long? = null,
        notes: String? = null
    ): Occupancy {
        return Occupancy(
            cropName = cropName,
            variety = variety,
            startDate = startDate,
            expectedHarvestDate = expectedHarvestDate,
            expectedReleaseDate = expectedReleaseDate,
            status = OccupancyStatus.ACTIVE.name,
            notes = notes,
            growingSpaceId = growingSpace?.id,
            plantId = plant?.id
        )
    }

    fun completeOccupancy(occupancy: Occupancy, endDate: Long? = null): Occupancy {
        return occupancy.copy(
            endDate = endDate ?: System.currentTimeMillis(),
            status = OccupancyStatus.COMPLETED.name,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun updateOccupancy(
        occupancy: Occupancy,
        cropName: String,
        variety: String?,
        startDate: Long,
        expectedHarvestDate: Long?,
        expectedReleaseDate: Long?,
        notes: String?
    ): Occupancy {
        return occupancy.copy(
            cropName = cropName,
            variety = variety,
            startDate = startDate,
            expectedHarvestDate = expectedHarvestDate,
            expectedReleaseDate = expectedReleaseDate,
            notes = notes,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun createDesire(
        cropName: String,
        variety: String? = null,
        notes: String? = null,
        garden: Garden? = null
    ): Desire {
        return Desire(
            cropName = cropName,
            variety = variety,
            notes = notes,
            gardenId = garden?.id
        )
    }

    fun updateDesire(
        desire: Desire,
        cropName: String,
        variety: String?,
        notes: String?
    ): Desire {
        return desire.copy(
            cropName = cropName,
            variety = variety,
            notes = notes,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun fulfillDesire(desire: Desire): Desire {
        return desire.copy(
            isFulfilled = true,
            isCancelled = false,
            isExpired = false,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun cancelDesire(desire: Desire): Desire {
        return desire.copy(
            isCancelled = true,
            isFulfilled = false,
            isExpired = false,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun expireDesire(desire: Desire): Desire {
        return desire.copy(
            isExpired = true,
            isFulfilled = false,
            isCancelled = false,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun createPlannedPlanting(
        cropName: String,
        variety: String? = null,
        plannedDate: Long? = null,
        growingSpace: GrowingSpace? = null,
        desire: Desire? = null,
        seed: Seed? = null,
        notes: String? = null,
        garden: Garden? = null
    ): PlannedPlanting {
        return PlannedPlanting(
            cropName = cropName,
            variety = variety,
            plannedDate = plannedDate,
            status = com.soilandsupper.domain.model.PlanStatus.PLANNED.name,
            notes = notes,
            gardenId = garden?.id,
            growingSpaceId = growingSpace?.id,
            desireId = desire?.id,
            seedId = seed?.id
        )
    }

    fun updatePlannedPlanting(
        plan: PlannedPlanting,
        cropName: String,
        variety: String?,
        plannedDate: Long?,
        growingSpace: GrowingSpace?,
        desire: Desire?,
        seed: Seed?,
        notes: String?
    ): PlannedPlanting {
        return plan.copy(
            cropName = cropName,
            variety = variety,
            plannedDate = plannedDate,
            notes = notes,
            growingSpaceId = growingSpace?.id,
            desireId = desire?.id,
            seedId = seed?.id,
            updatedAt = System.currentTimeMillis()
        )
    }

    fun cancelPlannedPlanting(plan: PlannedPlanting): PlannedPlanting {
        return plan.copy(
            status = com.soilandsupper.domain.model.PlanStatus.CANCELLED.name,
            updatedAt = System.currentTimeMillis()
        )
    }
}
